baybe/recommenders/pure/bayesian/botorch.py

```diff
@@ -268,26 +268,28 @@ def _recommend_hybrid(
             or None,  # TODO: https://github.com/pytorch/botorch/issues/2042
         )
 
-        disc_points = points[:, :num_comp_columns]
-        cont_points = points[:, num_comp_columns:]
-
-        # Get selected candidate indices
-        idxs = pd.Index(
-            pd.merge(
-                candidates_comp.reset_index(),
-                pd.DataFrame(disc_points, columns=candidates_comp.columns),
-                on=list(candidates_comp),
-            )["index"]
+        # Align candidates with search space index. Done via including the search space
+        # index during the merge, which is used later for back-translation into the
+        # experimental representation
+        merged = pd.merge(
+            pd.DataFrame(points),
+            candidates_comp.reset_index(),
+            on=list(candidates_comp.columns),
+            how="left",
+        ).set_index("index")
+
+        # Get experimental representation of discrete part
+        rec_disc_exp = searchspace.discrete.exp_rep.loc[merged.index]
+
+        # Combine discrete and continuous parts
+        rec_exp = pd.concat(
+            [
+                rec_disc_exp,
+                merged.iloc[:, num_comp_columns:].set_axis(
+                    searchspace.continuous.param_names, axis=1
+                ),
+            ],
+            axis=1,
```

Comment:
Sorry, haven't had the time yet to debug through the logic, but just conceptually:
* why don't we need to set the columns when creating the dataframe for the merge?! Seems very odd to me. Especially since columns are set in the merge for discrete spaces.
* Is it really still necessary to manually handle parts separately? I would have thought that by just merging all point dimensions there would be no more manual fiddling be required

Reply:
re 1
the column names are just `0, 1, 2` in both dataframes

re 2
is that possible in this logic? keep in mind that before the merge i dont know which expeirmental points I have, I only have their computational equivalents

Reply:
re1
But why? Isn't that a bit weird? Sorry perhaps for rushing here a bit, have no time for proper review at the moment. But I see this line `candidates_comp.columns = list(range(num_comp_columns))  # type: ignore`
which is ONLY executed inside the if statement. I'm currently not much up-to-date regarding the hybrid logic, but seems all a bit strange to me

re2
probably you're right. I'll let you do whatever you think is the best here, finish my last open PR, and then need to jump

Reply:
yeah im not super sure about that if statement why is it there, I think practically its prob never false?

but `pd.DataFrame(points)` creates the columns like said and `candidates_comp` gets them assigned explicitly like that

Reply:
I removed the if block. It doesnt seem possible to me. Also looking at `optimize_acqf_mixed`, `fixed_features_list` is not allowed to be `None`

Maybe it stems from a time where this `_recommend_hybrid` was also expected to dot he job for a purely continuous space, then candidates being emtpy would have been a possibility, but not really here

@AVHopp If you believe removing this if block is a grave error please let me know asap